Let `codegen_transmute_operand` just handle everything

scottmcm opened the pull request:
When combined with rust-lang/rust#143720, this means `rvalue_creates_operand` can just return `true` for *every* `Rvalue`.  (A future PR could consider removing it, though just letting it optimize out is fine for now.)

It's nicer anyway, IMHO, because it avoids needing the layout checks to be consistent in the two places, and thus is an overall reduction in code.  Plus it's a more helpful building block when used in other places this way.

(TBH, it probably would have been better to have it this way the whole time, but I clearly didn't understand `rvalue_creates_operand` when I originally wrote rust-lang/rust#109843.)

scottmcm committed:
Let `codegen_transmute_operand` just handle everything

When combined with 143720, this means `rvalue_creates_operand` can just return `true` for *every* `Rvalue`.  (A future PR could consider removing it, though just letting it optimize out is fine for now.)

It's nicer anyway, IMHO, because it avoids needing the layout checks to be consistent in the two places, and thus is an overall reduction in code.  Plus it's a more helpful building block when used in other places this way.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit af1b680c5519693f1be092bee9c2ecf58eabf52a with merge 1790c15cb9074d67a9a5edd4ea709796b02df1ff...
<!-- homu: {"type":"TryBuildStarted","head_sha":"af1b680c5519693f1be092bee9c2ecf58eabf52a","merge_sha":"1790c15cb9074d67a9a5edd4ea709796b02df1ff"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16243660326/job/45865806931)
Build commit: 1790c15cb9074d67a9a5edd4ea709796b02df1ff (`1790c15cb9074d67a9a5edd4ea709796b02df1ff`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/16243660326/job/45865806931"},"merge_sha":"1790c15cb9074d67a9a5edd4ea709796b02df1ff"} -->

rust-timer commented:
Queued 1790c15cb9074d67a9a5edd4ea709796b02df1ff with parent bfc046a4b8d6b57db02540182466e989a9b0fb40, future [comparison URL](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=1790c15cb9074d67a9a5edd4ea709796b02df1ff).
There are currently 2 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (1790c15cb9074d67a9a5edd4ea709796b02df1ff): [comparison URL](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=1790c15cb9074d67a9a5edd4ea709796b02df1ff&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


Benchmarking this pull request means it may be perf-sensitive – we'll automatically label it not fit for rolling up. You can override this, but we strongly advise not to, due to possible changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=1790c15cb9074d67a9a5edd4ea709796b02df1ff&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 0.4% | [0.4%, 0.5%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=1790c15cb9074d67a9a5edd4ea709796b02df1ff&stat=max-rss)
<details>
<summary>Results (secondary -4.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.0% | [-4.0%, -4.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=1790c15cb9074d67a9a5edd4ea709796b02df1ff&stat=cycles:u)
<details>
<summary>Results (primary -2.1%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-2.1%, -2.1%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.1% | [-2.1%, -2.1%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=bfc046a4b8d6b57db02540182466e989a9b0fb40&end=1790c15cb9074d67a9a5edd4ea709796b02df1ff&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 464.174s -> 465.333s (0.25%)
**Artifact size**: 374.70 MiB -> 374.75 MiB (0.01%)

scottmcm committed:
re-enable direct `bitcast`s for Int/Float vector transmutes (but not ones involving pointers)

bors commented:
:umbrella: The latest upstream changes (presumably #144249) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

saethlin commented on tests/codegen/intrinsics/transmute-simd.rs:
Why doesn't this test use `std::simd` or the `core::arch::x86_64` types?

saethlin commented on compiler/rustc_codegen_ssa/src/mir/rvalue.rs:
```diff
@@ -967,37 +1020,6 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
-                    // from fat pointers to vectors of u16. (See #143194 #110021 ...)
-                    (abi::BackendRepr::SimdVector { .. }, _) | (_, abi::BackendRepr::SimdVector { .. }) => false,
-                }
-            }
```
Is there a reason to keep this function then?

WaffleLapkin replied:
Removing the function makes sense to me ✅

scottmcm committed:
Remove `rvalue_creates_operand` entirely

Split to a separate commit to it could be reverted later if necessary, should we get new `Rvalue`s where we can't handle it this way.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

rust-timer commented:
Queued 1ee335ba59c92cc05eb39b94c4e542718519b762 with parent 4ff3fa01cbdd468851b1b859541ee1c648cde7de, future [comparison URL](https://perf.rust-lang.org/compare.html?start=4ff3fa01cbdd468851b1b859541ee1c648cde7de&end=1ee335ba59c92cc05eb39b94c4e542718519b762).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.9 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

bors commented:
:pushpin: Commit b7e025cfb64094d8672c752a9fffb12e9bc79567 has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin b7e025cfb64094d8672c752a9fffb12e9bc79567 -->
<!-- homu: {"type":"Approved","sha":"b7e025cfb64094d8672c752a9fffb12e9bc79567","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b7e025cfb64094d8672c752a9fffb12e9bc79567 with merge 283a0746a244a88503fed61844f44df925ccdbb6...
<!-- homu: {"type":"BuildStarted","head_sha":"b7e025cfb64094d8672c752a9fffb12e9bc79567","merge_sha":"283a0746a244a88503fed61844f44df925ccdbb6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16544620793/job/46793765053)
Approved by: WaffleLapkin
Pushing 283a0746a244a88503fed61844f44df925ccdbb6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"WaffleLapkin","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/16544620793/job/46793765053"},"merge_sha":"283a0746a244a88503fed61844f44df925ccdbb6"} -->

rust-timer commented:
Finished benchmarking commit (283a0746a244a88503fed61844f44df925ccdbb6): [comparison URL](https://perf.rust-lang.org/compare.html?start=ce5fdd7d42aba9a2925692e11af2bd39cf37798a&end=283a0746a244a88503fed61844f44df925ccdbb6&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ce5fdd7d42aba9a2925692e11af2bd39cf37798a&end=283a0746a244a88503fed61844f44df925ccdbb6&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ce5fdd7d42aba9a2925692e11af2bd39cf37798a&end=283a0746a244a88503fed61844f44df925ccdbb6&stat=max-rss)
<details>
<summary>Results (primary 4.5%, secondary 1.5%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 4.5%  | [3.2%, 5.8%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 4.0%  | [4.0%, 4.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.1%, -1.1%] | 1     |
| All ❌✅ (primary)                 | 4.5%  | [3.2%, 5.8%]   | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ce5fdd7d42aba9a2925692e11af2bd39cf37798a&end=283a0746a244a88503fed61844f44df925ccdbb6&stat=cycles:u)
<details>
<summary>Results (secondary -8.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -8.2% | [-8.2%, -8.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ce5fdd7d42aba9a2925692e11af2bd39cf37798a&end=283a0746a244a88503fed61844f44df925ccdbb6&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 466.456s -> 467.05s (0.13%)
**Artifact size**: 376.58 MiB -> 376.65 MiB (0.02%)
